E2e protocol ricardo

corrideat opened the pull request.

corrideat committed:
Move test data attributes into Cypress commands

corrideat committed:
Improvements to tests.

* Exclude some spec files from running in CI
* Use longer usernames to avoid collisions
* Fix bug in preSendCheck for chatroom

corrideat committed:
Re-enable tests

corrideat committed:
Merge branch 'master' into e2e-protocol-ricardo

corrideat committed:
Disable native worker temporarily to allow tests to pass

corrideat committed:
Add two tests to CI

corrideat committed:
Update test suites used

corrideat committed:
Add checks and suppress chatroom leave error

corrideat committed:
Move preSendChecks to actions. Fix: group-chat fails

corrideat committed:
Bugfixes. TODO: remove log, etc., DRY sorting in identity, fix chat-message and chat-scrolling tests

corrideat committed:
Bugfixes for pending messages; remove debug logging

corrideat committed:
Move General chatroom creation to side-effect, fix bug with sometimes not joining it

corrideat committed:
Add back accidentally removed tests

corrideat committed:
Merge branch 'e2e-protocol' into e2e-protocol-ricardo

taoeffect commented on test/cypress/support/commands.js:
```diff
@@ -346,20 +438,15 @@ Cypress.Commands.add('giAcceptGroupInvite', (invitationLink, {
     cy.giLogin(existingMemberUsername, { bypassUI })
 
     // NOTE: should wait until all pendingKeyShares are removed
-    cy.getByDT('app').then(([el]) => {
-      cy.get(el).should('have.attr', 'data-pending-key-shares', '')
-      cy.get(el).should('have.attr', 'data-pending-publish-events', '0')
-    })
```
Nicely done here with the DRY code! 👍

taoeffect commented on test/cypress/support/commands.js:
```diff
@@ -21,6 +22,48 @@ const getParamsFromInvitationLink = invitationLink => {
   }
 }
 
+// Util function to perform checks using SBP
+// The function takes a name (to register it as a Cypress command) and custom
+// check function that takes SBP as its first parameter. It registers event
+// handlers for various events that may change the Chelonia state and returns
+// a Promise that resolves once the check passes.
+const cySbpCheckCommand = (name, customCheckFn) => {
```
Very cool function 👍

corrideat committed:
Fix flow types

taoeffect commented on frontend/controller/actions/chatroom.js:
```diff
@@ -237,12 +237,62 @@ export default (sbp('sbp/selectors/register', {
+          // chatroom member
+          if (state.users?.[params.data.username]) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
The code already checked `params?.hooks?`, don't need to check again

taoeffect commented on frontend/controller/actions/chatroom.js:
```diff
@@ -237,12 +237,62 @@ export default (sbp('sbp/selectors/register', {
+        // chatroom member
+        if (!state.users?.[params.data.member]) return false
+        if (params?.hooks?.preSendCheck) {
+          return params?.hooks?.preSendCheck(msg, state)
```
Same here

taoeffect commented on frontend/controller/actions/group.js:
```diff
@@ -620,7 +599,20 @@ export default (sbp('sbp/selectors/register', {
+          // Don't send if the member has already been added
+          if (state.chatRooms[params.data.chatRoomID]?.users?.[username]?.status === PROFILE_STATUS.ACTIVE) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
Here too

taoeffect commented on frontend/controller/actions/group.js:
```diff
@@ -673,14 +665,37 @@ export default (sbp('sbp/selectors/register', {
+            // Don't send if the member has already been removed
+            if (state?.profiles?.[params.data.member]?.status !== PROFILE_STATUS.ACTIVE) return false
+            if (params?.hooks?.preSendCheck) {
+              return params?.hooks?.preSendCheck(msg, state)
```
Here three

taoeffect commented on frontend/controller/actions/group.js:
```diff
@@ -673,14 +665,37 @@ export default (sbp('sbp/selectors/register', {
+            // Don't send if we've already been removed
+            if (state?.profiles?.[rootState.loggedIn.username]?.status !== PROFILE_STATUS.ACTIVE) return false
+            if (params?.hooks?.preSendCheck) {
+              return params?.hooks?.preSendCheck(msg, state)
```
And here as well

taoeffect commented on frontend/controller/actions/group.js:
```diff
@@ -872,10 +887,40 @@ export default (sbp('sbp/selectors/register', {
+          // Don't send if the member isn't an active chatroom member
+          if (state.chatRooms[params.data.chatRoomID]?.users?.[params.data.member]?.status !== PROFILE_STATUS.ACTIVE) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
And everywhere else

taoeffect commented on frontend/controller/actions/group.js:
```diff
@@ -872,10 +887,40 @@ export default (sbp('sbp/selectors/register', {
+        preSendCheck (msg, state) {
+          if (state?.profiles?.[rootState.loggedIn.username]?.status === PROFILE_STATUS.ACTIVE) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
Like here

taoeffect commented on frontend/controller/actions/identity.js:
```diff
@@ -614,8 +643,50 @@ export default (sbp('sbp/selectors/register', {
+        preSendCheck (msg, state) {
+          if (has(state.chatRooms, params.data.contractID)) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
.

taoeffect commented on frontend/controller/actions/identity.js:
```diff
@@ -614,8 +643,50 @@ export default (sbp('sbp/selectors/register', {
+        preSendCheck (msg, state) {
+          if (has(state.groups, params.data.groupContractID)) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
.

taoeffect commented on frontend/controller/actions/identity.js:
```diff
@@ -614,8 +643,50 @@ export default (sbp('sbp/selectors/register', {
+        preSendCheck (msg, state) {
+          if (!has(state.groups, params.data.groupContractID)) return false
+          if (params?.hooks?.preSendCheck) {
+            return params?.hooks?.preSendCheck(msg, state)
```
.

taoeffect commented on frontend/model/contracts/shared/nativeNotification.js:
```diff
@@ -18,8 +18,13 @@ export async function requestNotificationPermission (force: boolean = false): Pr
 
-  if (Notification.permission === 'granted') {
-    await sbp('service-worker/setup-push-subscription')
+  // TODO FIXME: TEMPORARILY DISABLED AS IT CAUSES AN ERROR DURING THE TESTS
+  if (isNaN(1) && Notification.permission === 'granted') {
```
Can you fix this now? If this is something that needs to be disabled in the testing environment, that can probably be done by checking if we're inside of Cypress.

Some useful links:

- https://docs.cypress.io/guides/guides/module-api#cypressrun
- https://docs.cypress.io/guides/guides/environment-variables

(And then updating `grunt.registerTask('cypress'` in `Gruntfile.js`; we already have `process.env.CI`)

taoeffect commented on frontend/views/pages/PendingApproval.vue:
```diff
@@ -52,13 +57,24 @@ export default ({
+        if (contractID === this.ephemeral.groupIdWhenMounted) {
+          this.ephemeral.isJoining = false
+          // TODO: Remove when unmounted
+          sbp('okTurtles.events/off', handler)
```
`/off` takes an event as the first parameter.

taoeffect commented on shared/domains/chelonia/internals.js:
```diff
@@ -338,6 +340,12 @@ export default (sbp('sbp/selectors/register', {
         const newEntry = await sbp('chelonia/private/queueEvent', contractID, () => {
```
So, as a totally separate issue, I think we need to revisit this code:

```js
  'chelonia/queueInvocation': (contractID, sbpInvocation) => {
    // We maintain two queues, contractID, used for internal events (i.e.,
    // from chelonia) and public:contractID, used for operations that need to
    // be done when no internal operations are running (e.g., calls from a
    // contract that require to be done after a sync)
    // The following waits for the internal (contractID) queue to be finished
    // and then pushes the operation requested into the public queue
    // This ensures that all internal operations have finished before these
    // other selectors are called
    return sbp('chelonia/private/queueEvent', contractID, ['chelonia/private/noop']).then(() => sbp('chelonia/private/queueEvent', 'public:' + contractID, sbpInvocation))
  },
```

This comment doesn't make sense.

>     // We maintain two queues, contractID, used for internal events (i.e.,
>     // from chelonia) and public:contractID, used for operations that need to
>     // be done when no internal operations are running 

..

>     // This ensures that all internal operations have finished before these
>     // other selectors are called

These comments are wrong. In fact it allows the `sbpInvocation` to run at the same time as something is running on the internal queue.

I do not see an issue with directly pushing to the `contactID` queue. Now that we've stabilized the tests we should be able to test this theory.

taoeffect replied:
> I do not see an issue with directly pushing to the `contactID` queue. Now that we've stabilized the tests we should be able to test this theory.

After discussion, we reached the conclusion that although the comment needs to be fixed to be more accurate and accurately explain the reason for the different queue, we still might want to keep the different queue (because it might prevent developers from screwing up Chelonia's internal operations; give an example)

taoeffect reviewed (changes requested):
Nicely done @corrideat!

Left some comments and change requests, also a big one related to `'chelonia/queueInvocation'`

corrideat committed:
Address PR comments

corrideat committed:
Bugfix: Fix group membership sync issue, re-joining us automatically

corrideat committed:
Cleanup & move logic to internalSideEffectsStack (part 1)

corrideat committed:
Various bugfixes and cleaning up.

- Added catch blocks to all or most free-standing promises
- Removed validation of outgoing messages; outgoing messages now
  go through processMessage instead
- Made processMessage side-effect free as much as possible
- Removed extraneous logging
- Improved the tests to verify that keys are actually rotated
- Fixed a bug in the way the invite links were displayed.
- Other smaller changes

corrideat committed:
Harmonise ChatMain scroll position with master and make topOffset a computed property

corrideat committed:
Clean up and add comments

corrideat committed:
Fixes to chatroom scrolling

corrideat committed:
Add timeout to scrolling test

corrideat committed:
Moved onChatScroll to call flush and minor improvements

corrideat committed:
Clean up and comment code

corrideat committed:
Add onChatScroll method wrapper

corrideat committed:
Remove check for null or undefined

corrideat committed:
Fixed (removed) spurious error for outgoing OP_KEY_REQUEST; fixed the way to assign globalThis.crypto

corrideat committed:
Remove unnecessary preSendCheck uses and consolidate them in contracts

corrideat committed:
Fixes for pubsub when switching sessions (reset)

corrideat committed:
Improvements for legibility

corrideat committed:
Remove pending from root state

taoeffect commented on frontend/controller/actions/group.js:
```diff
@@ -283,21 +257,41 @@ export default (sbp('sbp/selectors/register', {
+      // while we were offline. In this case, we should *not* re-join
+      // automatically, even if we have a valid invitation secret and are
+      // technically able to. However, if the previous situation we *should*
+      // attempt to rejoin if the action was user-initiated.
```
Think there's a typo here:

> However, if the previous situation we *should* attempt to rejoin if the action was user-initiated.

Instead of "if the previous situation", do you mean "in the previous situation"? Or "if in the previous situation"? Maybe you can just rewrite that sentence?

corrideat committed:
Remove await from contracts identity.js and other small improvements

corrideat committed:
Remove await for actions in identity contract

corrideat committed:
Rewrite wait
